### PR TITLE
Fix FormData clone boundary inconsistency issue

### PR DESCRIPTION
**Pull request by helloliuyiming**
Resolves https://github.com/cfug/dio/issues/2303

## Fix boundary inconsistency in FormData.clone method

Description:

This PR addresses an issue where cloning FormData objects resulted in different boundaries each time. The problem was caused by allocating a new boundary for each clone without any boundary reuse logic.

Changes made:
- Added an `initBoundary` parameter to the FormData constructor.
- Implemented boundary reuse logic when cloning FormData objects.
- Updated documentation to explain how `initBoundary` affects `boundaryName`.

Technical details:
- The `FormData` constructor now accepts an optional `initBoundary` parameter.
- When `initBoundary` is provided, it overrides the `boundaryName` configuration.
- The cloning process now reuses the existing boundary when appropriate.

Impact:
This fix ensures consistent behavior when cloning FormData objects and improves efficiency by reusing boundaries where possible. It should resolve issues related to unexpected boundary changes during FormData manipulation.

Testing:
- Added unit tests to verify boundary consistency across cloned FormData objects.
- Manually tested with various use cases to ensure backwards compatibility.

Note to reviewers:
Please pay special attention to the boundary reuse logic and any potential side effects on existing FormData usage.

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### CODE REVIEW

**Review by AlexV525 — COMMENTED** [review]
Is it possible to directly update the `_boundary` after the construction, so we don't need an extra argument for the constructor which won't increase the complexity for usages?

**Comment by helloliuyiming** [comment]
> Is it possible to directly update the `_boundary` after the construction, so we don't need an extra argument for the constructor which won't increase the complexity for usages?

I apologize for the delayed response as I've been quite busy lately.

Regarding your suggested solution - I want to make sure I understand it correctly. Would this be what you had in mind?
```dart
  FormData({
    String? initBoundary,
    this.boundaryName = _boundaryName,
    this.camelCaseContentDisposition = false,
  }) {
    _init();
    if(initBoundary != null) {
      _boundary = initBoundary;
    }
  }
```

**Comment by AlexV525** [comment]
No. What I mean is:

```dart
// Convenience method to clone finalized FormData when retrying requests.
FormData clone() {
  final clone = FormData();
  clone._boundary = _boundary; // <-- Replace the boundary.
  clone.fields.addAll(fields);
  for (final file in files) {
    clone.files.add(MapEntry(file.key, file.value.clone()));
  }
  return clone;
}
```

**Comment by helloliuyiming** [comment]
Ah, I see. Your solution is indeed much better than what I proposed. Thank you for the guidance. I have submitted a new one implementing your suggested approach. Please review again, thanks.

**Comment by AlexV525** [comment]
Please add the CHANGELOG entry and update the PR checklist status.

**Comment by helloliuyiming** [comment]
done!

File: dio/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ See the [Migration Guide][] for the complete breaking changes list.**
 
 - Update comments and strings with `MultipartFile`.
 - Removes redundant warnings when composing request options on Web.
+- Fix boundary inconsistency in FormData.clone method
```
Comment (AlexV525) [inline dio/CHANGELOG.md:10]:
```suggestion
- Fixes boundary inconsistency in `FormData.clone()`.
```

**Review by AlexV525 — APPROVED** [review]
Thanks for the update. One last nit

**Pull request merged**